Demos 518 new column for applications

connor-parke-icf opened the pull request.

connor-parke-icf commented on client/src/components/table/columns/useDemonstrationColumns.tsx:
```diff
@@ -0,0 +1,210 @@
+        },
+      },
+    }),
+    columnHelper.display({
+      id: "viewDetails",
+      cell: ({ row }) => {
+        // Always link to the parent demonstration page
+        // If this row is an amendment or extension, use the parent id and add the correct query param
+        let demoId = row.original.id;
+        let queryParam = "";
+        if (row.original.type === "amendment" && row.original.parentId) {
+          demoId = row.original.parentId;
+          queryParam = "amendments=true";
+        } else if (row.original.type === "extension" && row.original.parentId) {
+          demoId = row.original.parentId;
+          queryParam = "extensions=true";
+        }
+        const href = queryParam
+          ? `/demonstrations/${demoId}?${queryParam}`
+          : `/demonstrations/${demoId}`;
+        const handleClick = () => {
+          window.location.href = href;
+        };
+        return (
+          <SecondaryButton
+            type="button"
+            size="small"
+            onClick={handleClick}
+            className="px-2 py-0 text-sm font-medium"
+          >
+            View
+          </SecondaryButton>
+        );
+      },
+    }),
```
modified to make sure the view button links to the correct tab on the demo details page.

connor-parke-icf commented on client/src/pages/DemonstrationDetail.tsx:
```diff
@@ -19,21 +15,49 @@ import { ExtensionTable } from "components/table/tables/ExtensionTable";
 
 export const DemonstrationDetail = () => {
   const { id } = useParams<{ id: string }>();
+  const location = useLocation();
```
sets the tab based on the query param, like 

/demonstrations/1?extension=true or /demonstrations/1?amendment=true

tresdonicf replied:
When are these query params populated? Do we still need to do this work for if somebody was looking to `ctrl + c` their current address and share something at their current tab or does this already exist?

tresdonicf commented on client/src/hooks/useDemonstration.ts:
```diff
@@ -13,25 +16,55 @@ import {
   UPDATE_DEMONSTRATION_MUTATION,
 } from "queries/demonstrationQueries";
 
+type Demonstration = ServerDemonstration & {
+  extensions: {
+    name: string;
+    projectOfficer: { fullName: string };
+    extensionStatus: { name: string };
+  }[];
+};
+
```
Awesome job with the union type! Maybe we should add a TODO here to just make sure it's revisted. At some point we'll be able to `ctrl + f` for all these `TODO`s and group them into themes and make some tickets

tresdonicf commented on client/src/components/table/columns/useDemonstrationColumns.tsx:
I don't think the `hook` pattern is quite what we're looking for here - it's alright that it's named as one, but it's moreso just a function (which is still an improvement imo). In react terms, "hooks" are typically things that will be used by multiple components whereas this will be primarily used to clean up a component (still a great thing to use it for) - just wanted to mention this distinction

tresdonicf commented on client/src/mock-data/demonstrationMocks.ts:
```diff
@@ -116,6 +116,25 @@ export const demonstrationMocks: MockedResponse[] = [
             state: { name: "Montana" },
             projectOfficer: { fullName: "John Doe" },
             users: [{ id: "1" }],
+            amendments: [
```
That's something I was wondering about as well - I'm working on a `TODO.md` file in #101 with an item like this in it to revisit this and make sure it's cleaned up.

I think we should declare a much slimmer group of mock data (1-3 entities of all the important concepts - demonstration, user, etc)  and when we're doing something like testing pagination, we just render that same record in a loop instead of declaring multiple records. We could do this in a new tab in our "Dev Mode" sidebar area called "Tables". That way we don't need to declare a bunch of these and we still get what we need out of testing pagination

connor-parke-icf commented:
Renamed useDemonstrationColumns component to not use the "hook" naming convention. There is likely more to do here to make it a bit smoother, but this should suffice for now. Added TODO comments to various places where we could revisit.
